File: artify-frontend/src/services/orderService.js
import api from './api';
import { dummyOrders, dummyCarts } from '../data/dummyData';

export const placeOrder = async (orderRequest) => {
  try {
    const response = await api.post('/orders', orderRequest);
    return response;
  } catch (error) {
    console.warn('Backend unavailable, falling back to dummy placeOrder simulation', error);
    const user = JSON.parse(localStorage.getItem('user'));
    if (!user) throw new Error('Not authenticated');
    
    // Simulating moving items from user's cart to order
    const cartIdx = dummyCarts.findIndex(c => c.userId === user.id);
    if (cartIdx === -1 || dummyCarts[cartIdx].items.length === 0) {
      throw new Error('Cart is empty');
    }
    
    const cart = dummyCarts[cartIdx];
    const orderItems = cart.items.map((item, idx) => ({
      id: idx + 1,
      orderId: dummyOrders.length + 1,
      productId: item.productId,
      quantity: item.quantity,
      priceAtPurchase: item.product.price,
      product: item.product,
    }));
    
    const newOrder = {
      id: dummyOrders.length + 1,
      userId: user.id,
      totalAmount: cart.totalAmount,
      status: 'PENDING',
      shippingAddress: orderRequest.shippingAddress,
      createdAt: new Date().toISOString(),
      items: orderItems,
    };
    
    dummyOrders.push(newOrder);
    
    // Clear cart
    cart.items = [];
    cart.totalAmount = 0;
    
    return {
      success: true,
      message: 'Order placed successfully (Simulated)',
      data: newOrder,
    };
  }
};

export const getOrders = async () => {
  try {
    const response = await api.get('/orders');
    return response;
  } catch (error) {
    console.warn('Backend unavailable, falling back to dummy getOrders simulation', error);
    const user = JSON.parse(localStorage.getItem('user'));
    if (!user) throw new Error('Not authenticated');
    
    const userOrders = dummyOrders.filter(o => o.userId === user.id);
    return userOrders;
  }
};

export const getOrderById = async (id) => {
  try {
    const response = await api.get(`/orders/${id}`);
    return response;
  } catch (error) {
    console.warn(`Backend unavailable, falling back to dummy getOrderById simulation for id ${id}`, error);
    const order = dummyOrders.find(o => o.id === Number(id));
    if (order) {
      return {
        success: true,
        message: 'Order retrieved successfully (Simulated)',
        data: order,
      };
    }
    throw new Error('Order not found');
  }
};

export const cancelOrder = async (id) => {
  try {
    const response = await api.put(`/orders/${id}/cancel`);
    return response;
  } catch (error) {
    console.warn(`Backend unavailable, falling back to dummy cancelOrder simulation for id ${id}`, error);
    const orderIdx = dummyOrders.findIndex(o => o.id === Number(id));
    if (orderIdx !== -1) {
      const order = dummyOrders[orderIdx];
      if (order.status === 'PENDING' || order.status === 'CONFIRMED') {
        order.status = 'CANCELLED';
        return {
          success: true,
          message: 'Order cancelled successfully (Simulated)',
          data: order,
        };
      }
      throw new Error('Order cannot be cancelled in its current state');
    }
    throw new Error('Order not found');
  }
};

export const trackOrder = async (id) => {
  try {
    const response = await api.get(`/orders/${id}/track`);
    return response;
  } catch (error) {
    console.warn(`Backend unavailable, falling back to dummy trackOrder simulation for id ${id}`, error);
    const order = dummyOrders.find(o => o.id === Number(id));
    if (order) {
      return {
        success: true,
        message: 'Order tracking info retrieved (Simulated)',
        data: order,
      };
    }
    throw new Error('Order not found');
  }
};

File: artify-backend/src/main/java/com/artify/service/OrderService.java
package com.artify.service;

import com.artify.dto.request.OrderRequest;
import com.artify.dto.response.OrderItemResponse;
import com.artify.dto.response.OrderResponse;
import com.artify.dto.response.ProductResponse;
import com.artify.exception.BadRequestException;
import com.artify.exception.ResourceNotFoundException;
import com.artify.exception.UnauthorizedException;
import com.artify.model.*;
import com.artify.model.enums.OrderStatus;
import com.artify.repository.CartRepository;
import com.artify.repository.OrderRepository;
import com.artify.repository.ProductRepository;
import com.artify.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class OrderService {

    private final OrderRepository orderRepository;
    private final CartRepository cartRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;

    @Transactional
    public OrderResponse placeOrder(Long userId, OrderRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User", "id", userId));

        Cart cart = cartRepository.findByUserId(userId)
                .orElseThrow(() -> new ResourceNotFoundException("Cart", "userId", userId));

        if (cart.getItems() == null || cart.getItems().isEmpty()) {
            throw new BadRequestException("Cart is empty. Cannot place order.");
        }

        Order order = Order.builder()
                .user(user)
                .shippingAddress(request.getShippingAddress())
                .status(OrderStatus.PENDING)
                .items(new ArrayList<>())
                .totalAmount(0.0)
                .build();

        double totalAmount = 0.0;

        for (CartItem cartItem : cart.getItems()) {
            Product product = cartItem.getProduct();

            if (product.getStock() < cartItem.getQuantity()) {
                throw new BadRequestException("Insufficient stock for product: " + product.getTitle());
            }

            OrderItem orderItem = OrderItem.builder()
                    .order(order)
                    .product(product)
                    .quantity(cartItem.getQuantity())
                    .priceAtPurchase(product.getPrice())
                    .build();

            order.getItems().add(orderItem);
            totalAmount += cartItem.getQuantity() * product.getPrice();

            product.setStock(product.getStock() - cartItem.getQuantity());
            productRepository.save(product);
        }

        order.setTotalAmount(totalAmount);
        order = orderRepository.save(order);

        cart.getItems().clear();
        cartRepository.save(cart);

        log.info("Order placed with id: {} for user: {}", order.getId(), userId);
        return mapToOrderResponse(order);
    }

    public List<OrderResponse> getOrders(Long userId) {
        return orderRepository.findByUserIdOrderByCreatedAtDesc(userId).stream()
                .map(this::mapToOrderResponse)
                .collect(Collectors.toList());
    }

    public OrderResponse getOrderById(Long orderId, Long userId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new ResourceNotFoundException("Order", "id", orderId));

        if (!order.getUser().getId().equals(userId)) {
            throw new UnauthorizedException("You are not authorized to view this order");
        }

        return mapToOrderResponse(order);
    }

    @Transactional
    public OrderResponse cancelOrder(Long orderId, Long userId) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new ResourceNotFoundException("Order", "id", orderId));

        if (!order.getUser().getId().equals(userId)) {
            throw new UnauthorizedException("You are not authorized to cancel this order");
        }

        if (order.getStatus() != OrderStatus.PENDING && order.getStatus() != OrderStatus.CONFIRMED) {
            throw new BadRequestException("Order cannot be cancelled. Current status: " + order.getStatus());
        }

        order.setStatus(OrderStatus.CANCELLED);

        for (OrderItem item : order.getItems()) {
            Product product = item.getProduct();
            product.setStock(product.getStock() + item.getQuantity());
            productRepository.save(product);
        }

        order = orderRepository.save(order);
        log.info("Order cancelled with id: {}", orderId);
        return mapToOrderResponse(order);
    }

    public OrderResponse trackOrder(Long orderId, Long userId) {
        return getOrderById(orderId, userId);
    }

    private OrderResponse mapToOrderResponse(Order order) {
        List<OrderItemResponse> itemResponses = order.getItems().stream()
                .map(this::mapToOrderItemResponse)
                .collect(Collectors.toList());

        return OrderResponse.builder()
                .id(order.getId())
                .userId(order.getUser().getId())
                .totalAmount(order.getTotalAmount())
                .status(order.getStatus().name())
                .shippingAddress(order.getShippingAddress())
                .createdAt(order.getCreatedAt())
                .items(itemResponses)
                .build();
    }

    private OrderItemResponse mapToOrderItemResponse(OrderItem item) {
        return OrderItemResponse.builder()
                .id(item.getId())
                .orderId(item.getOrder().getId())
                .productId(item.getProduct().getId())
                .quantity(item.getQuantity())
                .priceAtPurchase(item.getPriceAtPurchase())
                .product(mapToProductResponse(item.getProduct()))
                .build();
    }

    private ProductResponse mapToProductResponse(Product product) {
        return ProductResponse.builder()
                .id(product.getId())
                .artistId(product.getArtist().getId())
                .categoryId(product.getCategory() != null ? product.getCategory().getId() : null)
                .title(product.getTitle())
                .description(product.getDescription())
                .price(product.getPrice())
                .imageUrl(product.getImageUrl())
                .stock(product.getStock())
                .status(product.getStatus().name())
                .createdAt(product.getCreatedAt())
                .artistName(product.getArtist().getUser().getName())
                .categoryName(product.getCategory() != null ? product.getCategory().getName() : null)
                .build();
    }
}
